File: Java/libraries/recognizers-text-date-time/src/main/java/com/microsoft/recognizers/text/datetime/french/parsers/FrenchHolidayParserConfiguration.java
package com.microsoft.recognizers.text.datetime.french.parsers;

import com.google.common.collect.ImmutableMap;
import com.microsoft.recognizers.text.datetime.french.extractors.FrenchHolidayExtractorConfiguration;
import com.microsoft.recognizers.text.datetime.parsers.BaseHolidayParserConfiguration;
import com.microsoft.recognizers.text.datetime.resources.FrenchDateTime;
import com.microsoft.recognizers.text.datetime.utilities.DateUtil;
import java.time.LocalDateTime;
import java.util.Arrays;
import java.util.HashMap;
import java.util.Map;
import java.util.function.IntFunction;

public class FrenchHolidayParserConfiguration extends BaseHolidayParserConfiguration {

    public FrenchHolidayParserConfiguration() {

        setHolidayRegexList(FrenchHolidayExtractorConfiguration.HolidayRegexList);

        final HashMap<String, Iterable<String>> holidayNamesMap = new HashMap<>();
        for (final Map.Entry<String, String[]> entry : FrenchDateTime.HolidayNames.entrySet()) {
            if (entry.getValue() instanceof String[]) {
                holidayNamesMap.put(entry.getKey(), Arrays.asList(entry.getValue()));
            }
        }
        setHolidayNames(ImmutableMap.copyOf(holidayNamesMap));
    }

    private static LocalDateTime newYear(final int year) {
        return DateUtil.safeCreateFromMinValue(year, 1, 1);
    }

    private static LocalDateTime newYearEve(final int year) {
        return DateUtil.safeCreateFromMinValue(year, 12, 31);
    }

    private static LocalDateTime christmasDay(final int year) {
        return DateUtil.safeCreateFromMinValue(year, 12, 25);
    }

    private static LocalDateTime christmasEve(final int year) {
        return DateUtil.safeCreateFromMinValue(year, 12, 24);
    }

    private static LocalDateTime valentinesDay(final int year) {
        return DateUtil.safeCreateFromMinValue(year, 2, 14);
    }

    private static LocalDateTime whiteLoverDay(final int year) {
        return DateUtil.safeCreateFromMinValue(year, 3, 14);
    }

    private static LocalDateTime foolDay(final int year) {
        return DateUtil.safeCreateFromMinValue(year, 4, 1);
    }

    private static LocalDateTime girlsDay(final int year) {
        return DateUtil.safeCreateFromMinValue(year, 3, 7);
    }

    private static LocalDateTime treePlantDay(final int year) {
        return DateUtil.safeCreateFromMinValue(year, 3, 12);
    }

    private static LocalDateTime femaleDay(final int year) {
        return DateUtil.safeCreateFromMinValue(year, 3, 8);
    }

    private static LocalDateTime childrenDay(final int year) {
        return DateUtil.safeCreateFromMinValue(year, 6, 1);
    }

    private static LocalDateTime youthDay(final int year) {
        return DateUtil.safeCreateFromMinValue(year, 5, 4);
    }

    private static LocalDateTime teacherDay(final int year) {
        return DateUtil.safeCreateFromMinValue(year, 9, 10);
    }

    private static LocalDateTime singlesDay(final int year) {
        return DateUtil.safeCreateFromMinValue(year, 11, 11);
    }

    private static LocalDateTime maoBirthday(final int year) {
        return DateUtil.safeCreateFromMinValue(year, 12, 26);
    }

    private static LocalDateTime inaugurationDay(final int year) {
        return DateUtil.safeCreateFromMinValue(year, 1, 20);
    }

    private static LocalDateTime groundhogDay(final int year) {
        return DateUtil.safeCreateFromMinValue(year, 2, 2);
    }

    private static LocalDateTime stPatrickDay(final int year) {
        return DateUtil.safeCreateFromMinValue(year, 3, 17);
    }

    private static LocalDateTime stGeorgeDay(final int year) {
        return DateUtil.safeCreateFromMinValue(year, 4, 23);
    }

    private static LocalDateTime mayday(final int year) {
        return DateUtil.safeCreateFromMinValue(year, 5, 1);
    }

    private static LocalDateTime cincoDeMayoday(final int year) {
        return DateUtil.safeCreateFromMinValue(year, 5, 5);
    }

    private static LocalDateTime baptisteDay(final int year) {
        return DateUtil.safeCreateFromMinValue(year, 6, 24);
    }

    private static LocalDateTime usaIndependenceDay(final int year) {
        return DateUtil.safeCreateFromMinValue(year, 7, 4);
    }

    private static LocalDateTime bastilleDay(final int year) {
        return DateUtil.safeCreateFromMinValue(year, 7, 14);
    }

    private static LocalDateTime halloweenDay(final int year) {
        return DateUtil.safeCreateFromMinValue(year, 10, 31);
    }

    private static LocalDateTime allHallowDay(final int year) {
        return DateUtil.safeCreateFromMinValue(year, 11, 1);
    }

    private static LocalDateTime allSoulsday(final int year) {
        return DateUtil.safeCreateFromMinValue(year, 11, 2);
    }

    private static LocalDateTime guyFawkesDay(final int year) {
        return DateUtil.safeCreateFromMinValue(year, 11, 5);
    }

    private static LocalDateTime veteransday(final int year) {
        return DateUtil.safeCreateFromMinValue(year, 11, 11);
    }

    protected static LocalDateTime fathersDay(final int year) {
        return DateUtil.safeCreateFromMinValue(year, 6, 17);
    }

    protected static LocalDateTime mothersDay(final int year) {
        return DateUtil.safeCreateFromMinValue(year, 5, 27);
    }

    protected static LocalDateTime labourDay(final int year) {
        return DateUtil.safeCreateFromMinValue(year, 5, 1);
    }

    @Override
    public int getSwiftYear(final String text) {
        final String trimmedText = text.trim();
        int swift = -10;
        if (trimmedText.endsWith("prochain")) {
            // next - 'l'annee prochain'
            swift = 1;
        } else if (trimmedText.endsWith("dernier")) {
            // last - 'l'annee dernier'
            swift = -1;
        } else if (trimmedText.endsWith("cette")) {
            // this - 'cette annees'
            swift = 0;
        }

        return swift;
    }

    public String sanitizeHolidayToken(final String holiday) {
        return holiday
            .replaceAll(" ", "")
            .replaceAll("'", "");
    }

    protected HashMap<String, IntFunction<LocalDateTime>> initHolidayFuncs() {
        return new HashMap<String, IntFunction<LocalDateTime>>(super.initHolidayFuncs()) {{
                put("maosbirthday", FrenchHolidayParserConfiguration::maoBirthday);
                put("yuandan", FrenchHolidayParserConfiguration::newYear);
                put("teachersday", FrenchHolidayParserConfiguration::teacherDay);
                put("singleday", FrenchHolidayParserConfiguration::singlesDay);
                put("allsaintsday", FrenchHolidayParserConfiguration::halloweenDay);
                put("youthday", FrenchHolidayParserConfiguration::youthDay);
                put("childrenday", FrenchHolidayParserConfiguration::childrenDay);
                put("femaleday", FrenchHolidayParserConfiguration::femaleDay);
                put("treeplantingday", FrenchHolidayParserConfiguration::treePlantDay);
                put("arborday", FrenchHolidayParserConfiguration::treePlantDay);
                put("girlsday", FrenchHolidayParserConfiguration::girlsDay);
                put("whiteloverday", FrenchHolidayParserConfiguration::whiteLoverDay);
                put("loverday", FrenchHolidayParserConfiguration::valentinesDay);
                put("christmas", FrenchHolidayParserConfiguration::christmasDay);
                put("xmas", FrenchHolidayParserConfiguration::christmasDay);
                put("newyear", FrenchHolidayParserConfiguration::newYear);
                put("newyearday", FrenchHolidayParserConfiguration::newYear);
                put("newyearsday", FrenchHolidayParserConfiguration::newYear);
                put("inaugurationday", FrenchHolidayParserConfiguration::inaugurationDay);
                put("groundhougday", FrenchHolidayParserConfiguration::groundhogDay);
                put("valentinesday", FrenchHolidayParserConfiguration::valentinesDay);
                put("stpatrickday", FrenchHolidayParserConfiguration::stPatrickDay);
                put("aprilfools", FrenchHolidayParserConfiguration::foolDay);
                put("stgeorgeday", FrenchHolidayParserConfiguration::stGeorgeDay);
                put("mayday", FrenchHolidayParserConfiguration::mayday);
                put("cincodemayoday", FrenchHolidayParserConfiguration::cincoDeMayoday);
                put("baptisteday", FrenchHolidayParserConfiguration::baptisteDay);
                put("usindependenceday", FrenchHolidayParserConfiguration::usaIndependenceDay);
                put("independenceday", FrenchHolidayParserConfiguration::usaIndependenceDay);
                put("bastilleday", FrenchHolidayParserConfiguration::bastilleDay);
                put("halloweenday", FrenchHolidayParserConfiguration::halloweenDay);
                put("allhallowday", FrenchHolidayParserConfiguration::allHallowDay);
                put("allsoulsday", FrenchHolidayParserConfiguration::allSoulsday);
                put("guyfawkesday", FrenchHolidayParserConfiguration::guyFawkesDay);
                put("veteransday", FrenchHolidayParserConfiguration::veteransday);
                put("christmaseve", FrenchHolidayParserConfiguration::christmasEve);
                put("newyeareve", FrenchHolidayParserConfiguration::newYearEve);
                put("fathersday", FrenchHolidayParserConfiguration::fathersDay);
                put("mothersday", FrenchHolidayParserConfiguration::mothersDay);
                put("labourday", FrenchHolidayParserConfiguration::labourDay);
            }
        };
    }
}

File: .NET/Microsoft.Recognizers.Text.DateTime/French/Parsers/FrenchHolidayParserConfiguration.cs
﻿// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.

using System;
using System.Collections.Generic;
using System.Collections.Immutable;
using Microsoft.Recognizers.Definitions.French;
using DateObject = System.DateTime;

// TODO: Finish translating Dictionary below
namespace Microsoft.Recognizers.Text.DateTime.French
{
    public class FrenchHolidayParserConfiguration : BaseHolidayParserConfiguration
    {
        public FrenchHolidayParserConfiguration(IDateTimeOptionsConfiguration config)
            : base(config)
        {
            this.HolidayRegexList = FrenchHolidayExtractorConfiguration.HolidayRegexList;
            this.HolidayNames = DateTimeDefinitions.HolidayNames.ToImmutableDictionary();
        }

        public override int GetSwiftYear(string text)
        {
            var trimmedText = text.Trim();

            var swift = -10;

            // @TODO move hardcoded values to resources file

            if (trimmedText.EndsWith("prochain", StringComparison.Ordinal))
            {
                // next - 'l'annee prochain'
                swift = 1;
            }
            else if (trimmedText.EndsWith("dernier", StringComparison.Ordinal))
            {
                // last - 'l'annee dernier'
                swift = -1;
            }
            else if (trimmedText.StartsWith("cette", StringComparison.Ordinal))
            {
                // this - 'cette annees'
                swift = 0;
            }

            return swift;
        }

        public override string SanitizeHolidayToken(string holiday)
        {
            return holiday
                .Replace(" ", string.Empty)
                .Replace("'", string.Empty);
        }

        protected override IDictionary<string, Func<int, DateObject>> InitHolidayFuncs()
        {
            return new Dictionary<string, Func<int, DateObject>>(base.InitHolidayFuncs())
            {
                { "maosbirthday", MaoBirthday },
                { "yuandan", NewYear },
                { "teachersday", TeacherDay },
                { "singleday", SinglesDay },
                { "allsaintsday", HalloweenDay },
                { "youthday", YouthDay },
                { "childrenday", ChildrenDay },
                { "femaleday", FemaleDay },
                { "treeplantingday", TreePlantDay },
                { "arborday", TreePlantDay },
                { "girlsday", GirlsDay },
                { "whiteloverday", WhiteLoverDay },
                { "loverday", ValentinesDay },
                { "christmas", ChristmasDay },
                { "xmas", ChristmasDay },
                { "newyear", NewYear },
                { "newyearday", NewYear },
                { "newyearsday", NewYear },
                { "inaugurationday", InaugurationDay },
                { "groundhougday", GroundhogDay },
                { "valentinesday", ValentinesDay },
                { "stpatrickday", StPatrickDay },
                { "aprilfools", FoolDay },
                { "stgeorgeday", StGeorgeDay },
                { "mayday", Mayday },
                { "cincodemayoday", CincoDeMayoDay },
                { "baptisteday", BaptisteDay },
                { "usindependenceday", UsaIndependenceDay },
                { "independenceday", UsaIndependenceDay },
                { "bastilleday", BastilleDay },
                { "halloweenday", HalloweenDay },
                { "allhallowday", AllHallowsDay },
                { "allsoulsday", AllSoulsDay },
                { "guyfawkesday", GuyFawkesDay },
                { "veteransday", VeteransDay },
                { "christmaseve", ChristmasEve },
                { "newyeareve", NewYearEve },
                { "fathersday", FathersDay },
                { "mothersday", MothersDay },
                { "labourday", InternationalWorkersDay },
            };
        }

        private static DateObject NewYear(int year) => new DateObject(year, 1, 1);

        private static DateObject NewYearEve(int year) => new DateObject(year, 12, 31);

        private static DateObject ChristmasDay(int year) => new DateObject(year, 12, 25);

        private static DateObject ChristmasEve(int year) => new DateObject(year, 12, 24);

        private static DateObject ValentinesDay(int year) => new DateObject(year, 2, 14);

        private static DateObject WhiteLoverDay(int year) => new DateObject(year, 3, 14);

        private static DateObject FoolDay(int year) => new DateObject(year, 4, 1);

        private static DateObject GirlsDay(int year) => new DateObject(year, 3, 7);

        private static DateObject TreePlantDay(int year) => new DateObject(year, 3, 12);

        private static DateObject FemaleDay(int year) => new DateObject(year, 3, 8);

        private static DateObject ChildrenDay(int year) => new DateObject(year, 6, 1);

        private static DateObject YouthDay(int year) => new DateObject(year, 5, 4);

        private static DateObject TeacherDay(int year) => new DateObject(year, 9, 10);

        private static DateObject SinglesDay(int year) => new DateObject(year, 11, 11);

        private static DateObject MaoBirthday(int year) => new DateObject(year, 12, 26);

        private static DateObject InaugurationDay(int year) => new DateObject(year, 1, 20);

        private static DateObject GroundhogDay(int year) => new DateObject(year, 2, 2);

        private static DateObject StPatrickDay(int year) => new DateObject(year, 3, 17);

        private static DateObject StGeorgeDay(int year) => new DateObject(year, 4, 23);

        private static DateObject Mayday(int year) => new DateObject(year, 5, 1);

        private static DateObject CincoDeMayoDay(int year) => new DateObject(year, 5, 5);

        private static DateObject BaptisteDay(int year) => new DateObject(year, 6, 24);

        private static DateObject UsaIndependenceDay(int year) => new DateObject(year, 7, 4);

        private static DateObject BastilleDay(int year) => new DateObject(year, 7, 14);

        private static DateObject HalloweenDay(int year) => new DateObject(year, 10, 31);

        private static DateObject AllHallowsDay(int year) => new DateObject(year, 11, 1);

        private static DateObject AllSoulsDay(int year) => new DateObject(year, 11, 2);

        private static DateObject GuyFawkesDay(int year) => new DateObject(year, 11, 5);

        private static DateObject VeteransDay(int year) => new DateObject(year, 11, 11);

        private static new DateObject FathersDay(int year) => new DateObject(year, 6, 17);

        private static new DateObject MothersDay(int year) => new DateObject(year, 5, 27);

    }
}
